src/VisualStudio/Core/Def/Implementation/TableDataSource/Suppression/VisualStudioDiagnosticListSuppressionStateService.cs

```diff
@@ -185,28 +230,139 @@ public ImmutableArray<DiagnosticData> GetItems(bool selectedEntriesOnly, bool is
+                            // bail out
                             continue;
                         }
 
-                        if (isAddSuppression)
+                        Document document = null;
+                        var hasLocation = (entryHandle.TryGetValue(StandardTableColumnDefinitions.DocumentName, out filePath) && !string.IsNullOrEmpty(filePath)) &&
+                            (entryHandle.TryGetValue(StandardTableColumnDefinitions.Line, out line) && line >= 0);
+                        if (hasLocation)
                         {
-                            // Compiler diagnostics can only be suppressed in source.
-                            if (!diagnosticData.IsSuppressed &&
-                                (isSuppressionInSource || !isCompilerDiagnostic))
+                            if (string.IsNullOrEmpty(filePath) || line < 0)
+                            {
+                                // bail out
+                                continue;
+                            }
+
+                            ImmutableDictionary<string, Document> filePathMap;
+                            filePathToDocumentMapOpt = filePathToDocumentMapOpt ?? new Dictionary<Project, ImmutableDictionary<string, Document>>();
+                            if (!filePathToDocumentMapOpt.TryGetValue(project, out filePathMap))
                             {
-                                builder.Add(diagnosticData);
+                                filePathMap = await GetFilePathToDocumentMapAsync(project, cancellationToken).ConfigureAwait(false);
+                                filePathToDocumentMapOpt[project] = filePathMap;
                             }
+
+                            if (!filePathMap.TryGetValue(filePath, out document))
+                            {
+                                // bail out
+                                continue;
+                            }
+
+                            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                            var linePosition = new LinePosition(line, 0);
+                            var linePositionSpan = new LinePositionSpan(start: linePosition, end: linePosition);
+                            var textSpan = (await tree.GetTextAsync(cancellationToken).ConfigureAwait(false)).Lines.GetTextSpan(linePositionSpan);
+                            location = tree.GetLocation(textSpan);
                         }
-                        else if (diagnosticData.IsSuppressed)
+
+                        Contract.ThrowIfNull(project);
+                        Contract.ThrowIfFalse((document != null) == location.IsInSource);
+
+                        // Create a diagnostic with correct values for fields we care about: id, category, message, isSuppressed, location
+                        // and default values for the rest of the fields (not used by suppression fixer).
+                        var diagnostic = Diagnostic.Create(
```

Comment:
why create diagnostic and convert it to diagnosticdata rather than create diagnosticdata directly?